src/app/Fake.Tools.Octo/Octo.fs

```diff
@@ -0,0 +1,269 @@
+      (stringListParam "package" opts.Packages)
+      (optionalStringParam "packagesfolder" opts.PackagesFolder)
+      (optionalStringParam "releasenotes" (String.liftString opts.ReleaseNotes))
+      (optionalStringParam "releasenotesfile" (String.liftString opts.ReleaseNotesFile))
+      (flag "ignoreExisting" opts.IgnoreExisting)
+      (optionalStringParam "channel" opts.Channel)
+      (flag "ignorechannelrules" opts.IgnoreChannelRules) ] 
+    |> List.fold (+) ""
+
+/// [omit]
+let deployCommandLine (opts:DeployReleaseOptions) = 
+    [ (optionalStringParam "project" (String.liftString opts.Project))
+      (optionalStringParam "deployto" (String.liftString opts.DeployTo))
+      (optionalStringParam "version" (String.liftString opts.Version))
+      (flag "force" opts.Force)
+      (flag "waitfordeployment" opts.WaitForDeployment)
+      (flag "norawlog" opts.NoRawLog)
+      (flag "progress" opts.Progress)
+      (optionalObjParam "deploymenttimeout" opts.DeploymentTimeout)
+      (optionalObjParam "deploymentchecksleepcycle" opts.DeploymentCheckSleepCycle)
+      (optionalStringParam "specificmachines" opts.SpecificMachines) ] 
+    |> List.fold (+) ""
+
+/// [omit]
+let deleteCommandLine (opts:DeleteReleaseOptions) =
+    [ (optionalStringParam "project" (String.liftString opts.Project))
+      (optionalStringParam "minversion" (String.liftString opts.MinVersion))
+      (optionalStringParam "maxversion" (String.liftString opts.MaxVersion)) ] 
+    |> List.fold (+) ""
+
+/// [omit]
+let serverCommandLine (opts:OctoServerOptions) = 
+    [ (optionalStringParam "server" (String.liftString opts.Server))
+      (optionalStringParam "apikey" (String.liftString opts.ApiKey)) ] 
+    |> List.fold (+) ""
+
+/// [omit]
+let pushCommandLine (opts : PushOptions) =
+    [ stringListParam "package" opts.Packages
+      flag "replace-existing" opts.ReplaceExisting ]
+    |> List.fold (+) ""
+
+/// Maps a command to string input for the octopus tools cli.
+let commandLine command =
+    match command with
+    | CreateRelease (opts, None) ->
+        sprintf " create-release%s" (releaseCommandLine opts)
+    | CreateRelease (opts, Some (dopts)) ->
+        sprintf " create-release%s%s" (releaseCommandLine opts) (deployCommandLine dopts)
+    | DeployRelease opts ->
+        sprintf " deploy-release%s" (deployCommandLine opts)
+    | DeleteRelease opts ->
+        sprintf " delete-releases%s" (deleteCommandLine opts)
+    | ListEnvironments -> 
+        " list-environments"
+    | Push opts -> 
+        sprintf " push%s" (pushCommandLine opts)
+
+let serverCommandLineForTracing (opts: OctoServerOptions) = serverCommandLine { opts with ApiKey = "(Removed for security purposes)" }
+
+/// This task calls the Octo.exe CLI.
+/// See [Octopus-Tools](https://github.com/OctopusDeploy/Octopus-Tools) for more details.
+/// ## Parameters
+///
+///  - `setParams` - Function used to overwrite the OctoTools default parameters.
+let Octo setParams =
```

Comment:
We would choose a different name as currently the usage is `Octo.Octo` in fake 5 we opt for `Octo.run` or `Octo.exec` or whatever the correct verb is.

Reply:
The full list of Octo.exe verbs is [quite long](https://octopus.com/docs/api-and-integration/octo.exe-command-line). Shouldn't I create many verbs instead of just `run`:

```fsharp
let deployRelease setParams = ... 
let createRelease setParams = ...
```
Called like:
```fsharp
Octo.deployRelease (fun ps -> {ps with ...}
Octo.createRelease (fun ps -> {ps with ...}

```

Reply:
That depends on the representation of the `setParams` type. Both styles are valid according to the fake 5 guideline. I guess if the verbs are quite different then different functions with different parameter records is easier for users. If parameters are very equal across the different verbs than sharing (or only having a single function) makes more sense.

I'll leave that up to you as I'm not familiar with `octo.exe`.

Reply:
Thanks, there is only a handful of common options. Also the verbs do actually do quite different things. So I will go for separate verbs then.